docs/design/metainfo-rpc.md

```diff
@@ -0,0 +1,129 @@
+# Metainfo RPC Service Refactor
+
+## Background
+
+Our current metainfo rpc service isn't very future proof.
+If we shipped our current client code, it wouldn't be easy to update clients and add support for the following:
+- versions
+- multipart uploads
+- cleanup for uncommitted segments
+- general concurrency fixes
+
+Multipart upload can't happen with our current architecture because there's no concept of starting and ending a stream.
+
+Versions also wouldn't work, because if concurrent clients tried to upload to the same version, they would clobber each other.
+
+This should also reduce the number of roundtrips needed to start and end both streams and segments.
+
+This design should ensure compatibility between client and server in the future as these features are implemented. Neither clients nor servers will be able to access more information than what we provide in this new proto, so we need to make sure the messages and service calls anticipate needs for the above features.
+
+## Design
+
+The following is a proto file for just the rpc portion to create segments. The read portions just need to have a stream id included, which should default to being the largest stream id.
+(When we implement versions, every upload will have a new stream, which will be given the largest stream id at that time. This will allow us to get older versions by checking the stream id.)
+
+With this design, streams will be created that contain multiple segments. Previously, there wasn't a finalization step between segments, which meant that a big global object was created at a certain path. This prevented multipart uploads because the start/end of each segment was not tracked.
```

Comment:
I think this should read "streams will be created that contain multiple **parts**." There were already RPCs to finalize individual segments, but each segment was not logically associated with a part, but instead just the specific object.

Also, I don't remember the details, but I think part id is important to the server somehow (IIRC there's some finalization step).

Reply:
also added what you said here and took out the last sentence which I think was wrong